spoilers: Enhance toggle handling for improved click responsiveness

wulanseruniati opened the pull request:
This pull request improves the selection logic for spoiler headers in the Zulip app, enhancing the user experience when toggling spoilers.

Key Points Addressed:
Preventing Duplicate Toggles: The flag prevents toggle_spoiler from being called multiple times within a single interaction, ensuring that each click properly responds to the current state of the spoiler.
Early Exit for Selection: The logic to allow text selection without triggering a toggle ensures that clicks made during text selection don't inadvertently change the spoiler's state.
Improved User Experience: By managing how and when the toggle function is called, the code should provide a more consistent user experience, even during rapid interactions.

Fixes: [#30770](https://github.com/zulip/zulip/issues/30770)

[GIF of the new spoiler toggle functionality](https://ibb.co.com/54sjVGY)

<details>
<summary>Self-review checklist</summary>

- [ v ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
- [ v ] (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ v ] Explains differences from previous plans (e.g., issue description).
- [ v ] Highlights technical choices and bugs encountered.
- [ v ] Calls out remaining decisions and concerns.
- [ v ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ v ] Each commit is a coherent idea.
- [ v ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ v ] Visual appearance of the changes.
- [ v ] Responsiveness and internationalization.
- [ v ] Strings and tooltips.
- [ v ] End-to-end functionality of buttons, interactions and flows.
- [ v ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

There is also a problem with your uploaded GIF:

![Screenshot 2024-10-07 at 12 52 28@2x](https://github.com/user-attachments/assets/977b16e4-c6f6-4ad6-90d8-3eb702f08132)

wulanseruniati commented:
Hi @alya ,
I've cleaned up the commit history as requested, and I’ve also fixed the link to the GIF.
Could you please review the changes again?
Thank you!

wulanseruniati commented:
Thank you for the feedback @karlstolley !

I’ve identified the issue with the spoiler header text selection. The problem was related to the selection clearing and toggling the spoiler multiple times. In this update, I’ve added logic to ensure the spoiler only toggles once after the selection is cleared, preventing the glitchy behavior of it briefly opening and closing.

Could you please check if this resolves the issue on your end?

karlstolley commented:
These changes look like a good improvement, but please clean up your commit history to align with https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html

userAdityaa commented:
Hey @wulanseruniati, Is this still work in progress ? 
And change your `camel case`  -> `snake case` as zulip in most of the case follows snake case.

wulanseruniati commented:
hey @userAdityaa thanks for noticing,in the last commit i already change the `camel case` into `snake case`naming convention,
could you Please let me know if there’s anything else that needs adjustment, thankyou
